### DISCUSSION TITLE
AP2 on OpenWRT?

### DISCUSSION THREAD

**Discussion by quantonian** [mikebrady/shairport-sync#1340] [Q&A] [answered]
Has anyone successfully installed AP2 on OpenWRT? I've been able to install and run AP2 on Pi OS with no issues but am getting the following on OpenWRT before compiling nqptp:
```
root@fi03:/nqptp# autoreconf -fi
autom4te: error: need GNU m4 1.4 or later: m4
aclocal: error: /usr/bin/autom4te failed with exit status: 1
autoreconf: error: aclocal failed with exit status: 1
root@fi03:/nqptp# opkg install m4
Package m4 (1.4.19-1) installed in root is up to date.
```
The only suggested resolution I can find for this error is to use "brew" m4, but I'm not sure if that's an option for OpenWRT?

OpenWRT v21.02 on Pi 3 A+

Much appreciated for any assistance on this!

**Comment by quantonian** [comment]
Update: I've realized the root cause of this issue is that there is no "build-essential" package on openwrt so I'm missing all of the compiling packages (c++, gcc, make, etc.). I found what I believe is the equivalent for openwrt here: https://openwrt.org/docs/guide-developer/toolchain/using_the_sdk#load_package_lists

After downloading the SDK package I run out of disk space when uncompressing the file. I'm working on adding more space but I'm realizing this is getting way over my head at this point. Is there anyone out there that is using openwrt that has the skillset to compile nqptp and shairport-sync? I tried copying these binaries from pi os to openwrt but no dice. Any help on this would be greatly appreciated!!

**Reply by quantonian** [reply]
Another Update: Making progress after expanding the root file system and installing the openwrt SDK. I was able to successfully build nqptp but am getting the following message when running autoreconf for shairport-snc. I'm currently looking into this but wanted to see if anyone might have experience with this error. -thanks
```
root@OpenWrt:~/shairport-sync# autoreconf -ifv
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /usr/bin/autoconf --force
configure.ac:448: warning: The preprocessor macro `STDC_HEADERS' is obsolete.
configure.ac:448:   Except in unusual embedded environments, you can safely include all
configure.ac:448:   ISO C90 headers unconditionally.
configure.ac:27: error: possibly undefined macro: AC_MSG_ERROR
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: error: /usr/bin/autoconf failed with exit status: 1

```

**Comment by mikebrady** [comment]
Sorry to have missed all this. There is a rather old Shairport Sync package for OpenWRT. However, it is built by cross compilation (typically from a Ubuntu host) using the OpenWRT build system. I'm afraid it's been rather neglected in the last few years. 

The repository [Shairport Sync for OpenWRT](https://github.com/mikebrady/shairport-sync-for-openwrt) should give you a starting point for building more recent Shairport Syncs, including, hopefully, the AP2 version.

**Reply by quantonian** [reply]
Mike, thanks for the response. I was planning on updating my progress. Yes, the whole issue was trying to compile on openwrt os. I've recently installed debian os and have been using the instructions from your OpenWRT link after downloading the latest dev code but it compiles the legacy shairport-sync code for AP1. I was able to successfully compile and install and run on openwrt when I realized it was version 3.3.8

I think I figured it out, I need to change ```PKG_VERSION:=3.3.8``` to the latest dev version in the ```Makefile```. I'll update shortly and hopefully close this out. As always, thanks for your time!

**Reply by mikebrady** [reply]
Thanks for that. For AP2, you'll need to add extra libraries as well...

**Comment by quantonian** [comment, marked as answer]
Was finally able to get this to build and run successfully by following Mike's instructions for OpentWRT here: https://github.com/mikebrady/shairport-sync-for-openwrt

Additional config (for debian os):

Install additional libraries
```./scripts/feeds -f install libavahi alsa-lib libdaemon libpopt libsoxr alsa-utils htop libtool libconfig  avahi-daemon libavahi-client libplist libsodium uuid libgcrypt libffmpeg```

note that ```libffmpeg``` contains the av libraries ```libavutil libavcodec libavformat libasound2```

Makefile modifications:
```
PKG_VERSION:=4.1-devi-104-g84dd2dc4  #put latest dev build here

CONFIGURE_ARGS+= \
        --with-alsa \
        --with-avahi \
        --with-soxr \
        --with-libdaemon \
        --with-ssl=openssl \
        --with-airplay-2 \

define Package/shairport-sync
  $(Package/shairport-sync/Default)
   DEPENDS:= +libpthread +libopenssl +libavahi-client +alsa-lib +libdaemon +libsoxr +libpopt +libconfig +libstdcpp +libplist +libsodium +libgcrypt +libffmpeg-full +libuuid
endef
```

**Reply by mikebrady** [reply]
Good work -- well done!

**Reply by quantonian** [reply]
Thanks Mike, I appreciate all your effort to make this software available!

**Comment by jasonkolodziej** [comment]
For any other folks wanting @mikebrady's awesome sauce with Airplay 2. There has been a recent commit to OpenWRT packages repo.

https://github.com/openwrt/packages/commit/7391ca3d49ef92ca6510152bae0260fd4b0a738f has version 4.x with `--with-airplay-2` in Makefile for package.

hopefully the easiest way for most folks that don't want to go through a hour build process -- unless for specific target reasons 👍

**Reply by mikebrady** [reply]
Many thanks for this!

**Reply by mengw198901** [reply]
> For any other folks wanting @mikebrady's awesome sauce with Airplay 2. There has been a recent commit to OpenWRT packages repo.
> 
> [openwrt/packages@7391ca3](https://github.com/openwrt/packages/commit/7391ca3d49ef92ca6510152bae0260fd4b0a738f) has version 4.x with `--with-airplay-2` in Makefile for package.
> 
> hopefully the easiest way for most folks that don't want to go through a hour build process -- unless for specific target reasons 👍


For any other folks wanting @mikebrady's awesome sauce with Airplay 2. There has been a recent commit to OpenWRT packages repo.

openwrt/packages@7391ca3 has version 4.x with --with-airplay-2 in Makefile for package.

hopefully the easiest way for most folks that don't want to go through a hour build process -- unless for specific target reasons 👍

openwrt/packages@7391ca3

I used the Makefile compile the shairport-sycn.

I modified the version of Makefile to
include $(TOPDIR)/rules.mk

PKG_NAME:=shairport-sync
PKG_VERSION:=4.3.7
PKG_RELEASE:=1

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
PKG_SOURCE_URL:=https://codeload.github.com/mikebrady/shairport-sync/tar.gz/$(PKG_VERSION)?
PKG_HASH:=skip

PKG_MAINTAINER:=Ted Hess thess@kitschensync.net,
Mike Brady mikebrady@eircom.net
PKG_LICENSE:=MIT
PKG_LICENSE_FILES:=COPYING LICENSES shairport.c
If I use openwrt 23.05.5, the output file is

openwrt-ramips-mt76x8-mediatek_linkit-smart-7688-initramfs-kernel.bin
openwrt-ramips-mt76x8-mediatek_linkit-smart-7688.manifest
openwrt-ramips-mt76x8-mediatek_linkit-smart-7688-squashfs-sysupgrade.bin
shairport-sync-mbedtls_3.3.9-2_mipsel_24kc.ipk
shairport-sync-mini_3.3.9-2_mipsel_24kc.ipk
shairport-sync-openssl_3.3.9-2_mipsel_24kc.ipk

If I use openwrt 24.10.2, the output file is

openwrt-ramips-mt76x8-mediatek_linkit-smart-7688-initramfs-kernel.bin
openwrt-ramips-mt76x8-mediatek_linkit-smart-7688.manifest
openwrt-ramips-mt76x8-mediatek_linkit-smart-7688-squashfs-sysupgrade.bin
shairport-sync-mbedtls_4.3.2-r6_mipsel_24kc.ipk
shairport-sync-mini_4.3.2-r6_mipsel_24kc.ipk
shairport-sync-openssl_4.3.2-r6_mipsel_24kc.ipk

Why does it happen? The version is not 4.3.7.

@jasonkolodziej @mikebrady
